[macOS] Third-party video apps lack standardized caption customization for FCC compliance

Phinny01 opened the pull request:
#### e6c85cb486036022a925987ecfc05957ecf60cc8
<pre>
[macOS] Third-party video apps lack standardized caption customization for FCC compliance
<a href="https://bugs.webkit.org/show_bug.cgi?id=302891">https://bugs.webkit.org/show_bug.cgi?id=302891</a>
<a href="https://rdar.apple.com/165153469">rdar://165153469</a>

Reviewed by Jer Noble.

Implements the same factory pattern used on iOS to prefer AVKit&apos;s
AVLegibleMediaOptionsMenuController when available, falling back to custom AppKit implementation
when not.

- Implement the +menuController factory method in _WKCaptionStyleMenuControllerMac with runtime detection
- Created _WKCaptionStyleMenuControllerAVKitMac for AVKit integration
- Maintains backward compatibility and consistent cross-platform API

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::captionStyleMenuController):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerAVKitMac.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerAVKitMac.mm: Copied from Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm.
(-[_WKCaptionStyleMenuControllerAVKitMac init]):
(-[_WKCaptionStyleMenuControllerAVKitMac rebuildMenu]):
(-[_WKCaptionStyleMenuControllerAVKitMac isAncestorOf:]):
(-[_WKCaptionStyleMenuControllerAVKitMac captionStyleMenu]):
(-[_WKCaptionStyleMenuControllerAVKitMac legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(-[_WKCaptionStyleMenuControllerAVKitMac findAndDismissPopoverMenus]):
(-[_WKCaptionStyleMenuControllerAVKitMac searchForMenuInteractionsInWindow:]):
(-[_WKCaptionStyleMenuControllerAVKitMac menuWillOpen:]):
(-[_WKCaptionStyleMenuControllerAVKitMac menuDidClose:]):
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerMac.mm:
(+[WKCaptionStyleMenuController menuController]):
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController isAncestorOf:]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController menu]):
(-[WKCaptionStyleMenuController setMenu:]):
(-[WKCaptionStyleMenuController profileMenuItemSelected:]):
(-[WKCaptionStyleMenuController profileMenuItemHighlighted:]):
(-[WKCaptionStyleMenuController menuWillOpen:]):
(-[WKCaptionStyleMenuController menuDidClose:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ProfileIDSavingAndRestoration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ContextMenuDismissalSearch)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuRebuildingWithAVKit)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuAncestryCheck)):

Canonical link: <a href="https://commits.webkit.org/303770@main">https://commits.webkit.org/303770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d7f27a6a58dcaed8284ef216d0889387eabdc18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140991 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85483 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135305 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6450 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5801 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102070 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69492 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136382 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82866 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4444 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2043 "Found 4 new API test failures: TestWebKitAPI.CaptionPreferenceTests.CaptionStyleMenu, TestWebKitAPI.CaptionPreferenceTests.CaptionStyleMenuDelegate, TestWebKitAPI.CaptionPreferenceTests.CaptionStyleMenuSelect, TestWebKitAPI.CaptionPreferenceTests.CaptionStyleMenuHighlight (failure)") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113551 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37720 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143637 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5606 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38308 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110443 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5688 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110626 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28074 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4306 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115870 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59356 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5661 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34193 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5507 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69113 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5750 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5617 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerAVKitMac.mm:
```diff
@@ -0,0 +1,232 @@
+}
+
+- (RetainPtr<NSMenu>)convertUIMenuToNSMenu:(UIMenu *)uiMenu
+{
```
As we talked about offline, this seems unnecessary. Lets see if we can use the AVLegibleMediaOptionsMenu in a way that just uses NSMenus directly.

Commit:
[macOS] Third-party video apps lack standardized caption customization for FCC compliance
https://bugs.webkit.org/show_bug.cgi?id=302891
rdar://165153469

Reviewed by Jer Noble.

Implements the same factory pattern used on iOS to prefer AVKit's
AVLegibleMediaOptionsMenuController when available, falling back to custom AppKit implementation
when not.

- Implement the +menuController factory method in _WKCaptionStyleMenuControllerMac with runtime detection
- Created _WKCaptionStyleMenuControllerAVKitMac for AVKit integration
- Maintains backward compatibility and consistent cross-platform API

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::captionStyleMenuController):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerAVKitMac.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerAVKitMac.mm: Copied from Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm.
(-[_WKCaptionStyleMenuControllerAVKitMac init]):
(-[_WKCaptionStyleMenuControllerAVKitMac rebuildMenu]):
(-[_WKCaptionStyleMenuControllerAVKitMac isAncestorOf:]):
(-[_WKCaptionStyleMenuControllerAVKitMac captionStyleMenu]):
(-[_WKCaptionStyleMenuControllerAVKitMac legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(-[_WKCaptionStyleMenuControllerAVKitMac findAndDismissPopoverMenus]):
(-[_WKCaptionStyleMenuControllerAVKitMac searchForMenuInteractionsInWindow:]):
(-[_WKCaptionStyleMenuControllerAVKitMac menuWillOpen:]):
(-[_WKCaptionStyleMenuControllerAVKitMac menuDidClose:]):
* Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuControllerMac.mm:
(+[WKCaptionStyleMenuController menuController]):
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController isAncestorOf:]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController menu]):
(-[WKCaptionStyleMenuController setMenu:]):
(-[WKCaptionStyleMenuController profileMenuItemSelected:]):
(-[WKCaptionStyleMenuController profileMenuItemHighlighted:]):
(-[WKCaptionStyleMenuController menuWillOpen:]):
(-[WKCaptionStyleMenuController menuDidClose:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ProfileIDSavingAndRestoration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ContextMenuDismissalSearch)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuRebuildingWithAVKit)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuAncestryCheck)):

Canonical link: https://commits.webkit.org/303770@main
